File: cpp/src/micro_pipeline_bench.cpp
/// Real batch pipeline micro benchmark — incrementally adds layers to find where C++ loses to Rust.
///
/// Case A: hashmap probe only (trivial init/update, no StoreKey, no compare)
/// Case B: hashmap + real GetUnequalsNumWithDecode batch compare
/// Case C: hashmap + batch compare + NewRow (no varchar serialize)
/// Case D: hashmap + batch compare + NewRow + StoreKeyOneRow (full pipeline)
///
/// Usage: ./cpp_micro_pipeline <sel> [iterations]
/// Default: sel=0.1, 10 iterations

#include <cstdio>
#include <cstdint>
#include <cstring>
#include <cstdlib>
#include <chrono>
#include <vector>
#include <string>
#include <numeric>
#include <algorithm>
#include <random>

#define XXH_INLINE_ALL
#include "xxhash.h"
#include "column_marshaller.h"

static constexpr size_t NUM_STR_COLS = 4;
static constexpr size_t NUM_INT_COLS = 0;
static constexpr size_t HT_SIZE = 16384;
static constexpr double LOAD_FACTOR = 0.50;
static constexpr size_t NUM_PROBE_ROWS = 1000000;
static constexpr size_t BATCH_SIZE = 410;
static constexpr uint64_t SEED = 42;

using Clock = std::chrono::high_resolution_clock;
static double elapsed_ms(Clock::time_point t0, Clock::time_point t1) {
    return std::chrono::duration<double, std::milli>(t1 - t0).count();
}

// ─── Data gen (same as profile_taper_standalone) ─────────────────────────
static inline uint64_t HB(const uint8_t* d, size_t l, uint64_t s) { return XXH3_64bits_withSeed(d, l, s); }

struct BenchData {
    std::vector<std::vector<std::vector<uint8_t>>> strCols;
    std::vector<int64_t> hashes, values;
    std::vector<std::vector<taper::VarcharSlice>> strSlices;
    size_t totalRows;
};

static BenchData GenData(size_t nKeys, double sel) {
    std::mt19937_64 rng(SEED);
    BenchData d;
    d.strCols.resize(NUM_STR_COLS);
    for (size_t c = 0; c < NUM_STR_COLS; c++) {
        d.strCols[c].resize(nKeys);
        for (size_t i = 0; i < nKeys; i++) {
            auto s = "key_" + std::to_string(i) + "_c" + std::to_string(c);
            d.strCols[c][i].assign(s.begin(), s.end());
        }
    }
    std::vector<int64_t> bh(nKeys);
    for (size_t i = 0; i < nKeys; i++) {
        uint64_t h = 0;
        for (size_t c = 0; c < NUM_STR_COLS; c++)
            h = HB(d.strCols[c][i].data(), d.strCols[c][i].size(), h);
        bh[i] = static_cast<int64_t>(h);
    }
    std::vector<int64_t> bv(nKeys);
    for (size_t i = 0; i < nKeys; i++) bv[i] = i % 1000;

    size_t nH = static_cast<size_t>(NUM_PROBE_ROWS * sel), nM = NUM_PROBE_ROWS - nH;
    std::vector<std::vector<std::vector<uint8_t>>> ps(NUM_STR_COLS);
    for (auto& v : ps) v.reserve(NUM_PROBE_ROWS);
    std::vector<int64_t> ph; ph.reserve(NUM_PROBE_ROWS);
    for (size_t i = 0; i < nH; i++) { size_t idx = rng() % nKeys; for (size_t c = 0; c < NUM_STR_COLS; c++) ps[c].push_back(d.strCols[c][idx]); ph.push_back(bh[idx]); }
    for (size_t i = 0; i < nM; i++) { uint64_t h = 0; for (size_t c = 0; c < NUM_STR_COLS; c++) { auto s = "miss_" + std::to_string(i) + "_" + std::to_string(c); std::vector<uint8_t> b(s.begin(), s.end()); h = HB(b.data(), b.size(), h); ps[c].push_back(std::move(b)); } ph.push_back(static_cast<int64_t>(h)); }
    std::vector<size_t> ord(NUM_PROBE_ROWS); std::iota(ord.begin(), ord.end(), 0);
    for (size_t i = NUM_PROBE_ROWS - 1; i > 0; i--) std::swap(ord[i], ord[rng() % (i + 1)]);
    for (size_t c = 0; c < NUM_STR_COLS; c++) { auto tmp = std::move(ps[c]); ps[c].resize(NUM_PROBE_ROWS); for (size_t i = 0; i < NUM_PROBE_ROWS; i++) ps[c][i] = std::move(tmp[ord[i]]); }
    { auto tmp = ph; for (size_t i = 0; i < NUM_PROBE_ROWS; i++) ph[i] = tmp[ord[i]]; }
    std::vector<int64_t> pv(NUM_PROBE_ROWS); for (size_t i = 0; i < NUM_PROBE_ROWS; i++) pv[i] = i % 1000;

    d.totalRows = nKeys + NUM_PROBE_ROWS;
    for (size_t c = 0; c < NUM_STR_COLS; c++) { d.strCols[c].reserve(d.totalRows); for (auto& v : ps[c]) d.strCols[c].push_back(std::move(v)); }
    d.hashes = bh; d.hashes.insert(d.hashes.end(), ph.begin(), ph.end());
    d.values = bv; d.values.insert(d.values.end(), pv.begin(), pv.end());
    d.strSlices.resize(NUM_STR_COLS);
    for (size_t c = 0; c < NUM_STR_COLS; c++) { d.strSlices[c].resize(d.totalRows); for (size_t i = 0; i < d.totalRows; i++) { d.strSlices[c][i].ptr = d.strCols[c][i].data(); d.strSlices[c][i].len = d.strCols[c][i].size(); } }
    return d;
}

// ─── Case C2: HashMap + NewRow + StoreKeyOneRow (serialize, no batch compare) ─
__attribute__((noinline, flatten))
static size_t RunHashmapPlusSerialize(const BenchData& d, size_t numChunks) {
    taper::SimpleArenaAllocator pool;
    std::vector<size_t> keySizes(4, 0);
    std::vector<taper::ColumnKind> kinds(4, taper::ColumnKind::Varchar);
    taper::RowContainer rc(keySizes, kinds, 8, pool);
    taper::TaperFlatHashTable table(numChunks);
    int32_t aggOffset = rc.AggStateOffset();
    uint64_t sum = 0;

    // Process in batches (same as real pipeline)
    size_t numBatches = (d.totalRows + BATCH_SIZE - 1) / BATCH_SIZE;
    std::vector<taper::ColumnInput> cols(NUM_STR_COLS);

    for (size_t batch = 0; batch < numBatches; batch++) {
        size_t start = batch * BATCH_SIZE;
        size_t end = std::min(start + BATCH_SIZE, d.totalRows);
        int32_t batchLen = static_cast<int32_t>(end - start);

        for (size_t c = 0; c < NUM_STR_COLS; c++)
            cols[c] = taper::ColumnInput::MakeVarchar(d.strSlices[c].data() + start);

        const taper::ColumnInput* colsPtr = cols.data();
        taper::RowContainer* rcPtr = &rc;

        table.EmplaceBatch(d.hashes.data() + start, batchLen,
            [](int32_t) { return false; },
            [&](uint32_t rowIdx, char* data) {
                char* row = rcPtr->NewRow();
                // SetRowPtr
                uint64_t ptr = reinterpret_cast<uint64_t>(row);
                memcpy(data, &ptr, taper::ROW_PTR_SIZE);
                // StoreKeyOneRow — serialize 4 varchars into merged block
                size_t totalSize = 0;
                for (size_t c = 0; c < NUM_STR_COLS; c++) {
                    totalSize += 1 + taper::ComputeRowLenSize(colsPtr[c].vcSlices[rowIdx].len) + colsPtr[c].vcSlices[rowIdx].len;
                }
                uint8_t* block = rcPtr->ArenaAlloc(totalSize);
                uint8_t* wp = block;
                for (size_t c = 0; c < NUM_STR_COLS; c++) {
                    wp += taper::SerializeVarcharToBuffer(wp, colsPtr[c].vcSlices[rowIdx].ptr, colsPtr[c].vcSlices[rowIdx].len);
                }
                memcpy(row, &block, sizeof(block)); // store pointer at offset 0 (varchar slot col)
                taper::RowContainer::StoreValue<int64_t>(row, aggOffset, d.values[start + rowIdx]);
            },
            [&sum](uint32_t, char*, bool isNew) { if (!isNew) sum++; }
        );
    }
    return table.Size() + sum;
}

// ─── Case D: Full pipeline (same as profile_taper_standalone) ─────────────
__attribute__((noinline))
static size_t RunFullPipeline(const BenchData& d, size_t numChunks) {
    taper::SimpleArenaAllocator pool;
    std::vector<taper::ColumnDesc> cd(NUM_STR_COLS, taper::ColumnDesc::Varchar);
    taper::TaperColumnSerializeHandler t(pool, 8, cd, numChunks);
    size_t numBatches = (d.totalRows + BATCH_SIZE - 1) / BATCH_SIZE;
    std::vector<taper::ColumnInput> cols(NUM_STR_COLS);
    for (size_t batch = 0; batch < numBatches; batch++) {
        size_t start = batch * BATCH_SIZE;
        size_t end = std::min(start + BATCH_SIZE, d.totalRows);
        int32_t batchLen = static_cast<int32_t>(end - start);
        for (size_t c = 0; c < NUM_STR_COLS; c++)
            cols[c] = taper::ColumnInput::MakeVarchar(d.strSlices[c].data() + start);
        t.EmplaceTableWithDecode(d.hashes.data() + start, batchLen, cols, d.values.data() + start);
    }
    return t.NumGroups();
}

// ─── Case A: HashMap probe only (trivial callbacks) ───────────────────────
__attribute__((noinline))
static size_t RunHashmapOnly(const BenchData& d, size_t numChunks) {
    taper::TaperFlatHashTable table(numChunks);
    uint64_t sum = 0;
    table.EmplaceBatch(d.hashes.data(), static_cast<int32_t>(d.totalRows),
        [](int32_t) { return false; },
        [](uint32_t, char* data) { memset(data, 0x42, taper::ROW_PTR_SIZE); },
        [&sum](uint32_t, char*, bool isNew) { if (!isNew) sum++; }
    );
    return table.Size() + sum;
}

// ─── Case C: HashMap + NewRow (no serialize) ──────────────────────────────
__attribute__((noinline))
static size_t RunHashmapPlusNewRow(const BenchData& d, size_t numChunks) {
    taper::SimpleArenaAllocator pool;
    std::vector<size_t> keySizes(4, 0);
    std::vector<taper::ColumnKind> kinds(4, taper::ColumnKind::Varchar);
    taper::RowContainer rc(keySizes, kinds, 8, pool);
    taper::TaperFlatHashTable table(numChunks);
    int32_t aggOffset = rc.AggStateOffset();
    uint64_t sum = 0;

    table.EmplaceBatch(d.hashes.data(), static_cast<int32_t>(d.totalRows),
        [](int32_t) { return false; },
        [&](uint32_t rowIdx, char* data) {
            char* row = rc.NewRow();
            // Just write pointer + agg, no varchar serialize
            uint64_t ptr = reinterpret_cast<uint64_t>(row);
            memcpy(data, &ptr, taper::ROW_PTR_SIZE);
            taper::RowContainer::StoreValue<int64_t>(row, aggOffset, d.values[rowIdx]);
        },
        [&sum](uint32_t, char*, bool isNew) { if (!isNew) sum++; }
    );
    return table.Size() + sum;
}

// ═══════════════════════════════════════════════════════════════════
int main(int argc, char** argv) {
    double sel = 0.1;
    size_t numIters = 10;
    const char* stage = nullptr; // null = run all

    for (int i = 1; i < argc; i++) {
        if (i == 1) sel = std::atof(argv[i]);
        else if (i == 2) {
            // Check if it's a stage name or iteration count
            if (argv[i][0] >= 'A' && argv[i][0] <= 'Z') {
                stage = argv[i];
            } else {
                numIters = static_cast<size_t>(std::atoi(argv[i]));
            }
        }
        else if (i == 3) numIters = static_cast<size_t>(std::atoi(argv[i]));
    }

    size_t numKeys = static_cast<size_t>(HT_SIZE * LOAD_FACTOR);
    size_t numMisses = NUM_PROBE_ROWS - static_cast<size_t>(NUM_PROBE_ROWS * sel);
    size_t distinctKeys = numKeys + numMisses;
    size_t minSlots = std::max(static_cast<size_t>(distinctKeys / 0.85), size_t(8));
    size_t numChunks = 1;
    while (numChunks * 8 < minSlots) numChunks *= 2;

    fprintf(stderr, "=== C++ Pipeline Micro Bench ===\n");
    fprintf(stderr, "sel=%.1f, iters=%zu, totalRows=%zu, numChunks=%zu\n", sel, numIters, numKeys + NUM_PROBE_ROWS, numChunks);
    if (stage) fprintf(stderr, "Stage filter: %s\n", stage);
    fprintf(stderr, "\nGenerating data...\n");
    BenchData data = GenData(numKeys, sel);
    fprintf(stderr, "Done.\n\n");

    auto bench = [&](const char* name, const char* tag, size_t(*fn)(const BenchData&, size_t)) {
        if (stage && strcmp(stage, tag) != 0) return;
        volatile size_t w = fn(data, numChunks); (void)w;
        auto t0 = Clock::now();
        volatile size_t result = 0;
        for (size_t i = 0; i < numIters; i++) result = fn(data, numChunks);
        auto t1 = Clock::now();
        double per_iter = elapsed_ms(t0, t1) / numIters;
        printf("%-30s  per_iter=%7.2f ms  result=%zu\n", name, per_iter, (size_t)result);
    };

    printf("=== Results (sel=%.1f, %zu iters) ===\n", sel, numIters);
    bench("A: hashmap_only", "A", RunHashmapOnly);
    bench("C: hashmap+newrow", "C", RunHashmapPlusNewRow);
    bench("C2: hashmap+newrow+serialize", "C2", RunHashmapPlusSerialize);
    bench("D: full_pipeline", "D", RunFullPipeline);

    return 0;
}

File: rust/src/bin/micro_pipeline_bench.rs
//! Real batch pipeline micro benchmark — incrementally adds layers.
//!
//! Case A: hashmap probe only
//! Case C: hashmap + NewRow (no varchar serialize)
//! Case D: full pipeline (EmplaceTableWithDecode)
//!
//! Usage: ./micro_pipeline_bench <sel> [iterations]

use std::time::Instant;
use taper_hashmap::column_marshaller::{TaperColumnSerializeHandler, ColumnDesc, ColumnInput};
use taper_hashmap::taper_hashmap::TaperHashMap;
use taper_hashmap::chunk::SlotValue;
use taper_hashmap::row_container::{RowContainer, ColumnKind};
use xxhash_rust::xxh3::xxh3_64_with_seed;
use rand_mt::Mt19937GenRand64;

const NUM_STR_COLS: usize = 4;
const HT_SIZE: usize = 16384;
const LOAD_FACTOR: f64 = 0.50;
const NUM_PROBE_ROWS: usize = 1_000_000;
const BATCH_SIZE: usize = 410;
const SEED: u64 = 42;

fn hash_bytes(data: &[u8], seed: u64) -> u64 { xxh3_64_with_seed(data, seed) }
fn gen_string(base: &str, id: usize, col: usize) -> Vec<u8> { format!("{}_{}_c{}", base, id, col).into_bytes() }

struct BenchData {
    str_cols: Vec<Vec<Vec<u8>>>,
    hashes: Vec<u64>,
    values: Vec<i64>,
    total_rows: usize,
}

fn gen_data(num_keys: usize, sel: f64) -> BenchData {
    let mut rng = Mt19937GenRand64::new(SEED);
    let mut str_cols: Vec<Vec<Vec<u8>>> = (0..NUM_STR_COLS)
        .map(|c| (0..num_keys).map(|i| gen_string("key", i, c)).collect()).collect();
    let build_hashes: Vec<u64> = (0..num_keys).map(|i| {
        let mut h = 0u64;
        for c in 0..NUM_STR_COLS { h = hash_bytes(&str_cols[c][i], h); }
        h
    }).collect();
    let build_values: Vec<i64> = (0..num_keys).map(|i| (i % 1000) as i64).collect();

    let num_hits = (NUM_PROBE_ROWS as f64 * sel) as usize;
    let num_misses = NUM_PROBE_ROWS - num_hits;
    let mut probe_str_cols: Vec<Vec<Vec<u8>>> = vec![Vec::with_capacity(NUM_PROBE_ROWS); NUM_STR_COLS];
    let mut probe_hashes: Vec<u64> = Vec::with_capacity(NUM_PROBE_ROWS);
    for _ in 0..num_hits { let idx = (rng.next_u64() as usize) % num_keys; for c in 0..NUM_STR_COLS { probe_str_cols[c].push(str_cols[c][idx].clone()); } probe_hashes.push(build_hashes[idx]); }
    for i in 0..num_misses { let mut h = 0u64; for c in 0..NUM_STR_COLS { let s = format!("miss_{}_{}", i, c).into_bytes(); h = hash_bytes(&s, h); probe_str_cols[c].push(s); } probe_hashes.push(h); }

    let mut order: Vec<usize> = (0..NUM_PROBE_ROWS).collect();
    for i in (1..NUM_PROBE_ROWS).rev() { order.swap(i, (rng.next_u64() as usize) % (i + 1)); }
    let probe_str_cols: Vec<Vec<Vec<u8>>> = (0..NUM_STR_COLS).map(|c| order.iter().map(|&i| probe_str_cols[c][i].clone()).collect()).collect();
    let probe_hashes: Vec<u64> = order.iter().map(|&i| probe_hashes[i]).collect();
    let probe_values: Vec<i64> = (0..NUM_PROBE_ROWS).map(|i| (i % 1000) as i64).collect();

    let total_rows = num_keys + NUM_PROBE_ROWS;
    for c in 0..NUM_STR_COLS { str_cols[c].extend(probe_str_cols[c].iter().cloned()); }
    let mut all_hashes = build_hashes; all_hashes.extend_from_slice(&probe_hashes);
    let mut all_values = build_values; all_values.extend_from_slice(&probe_values);

    BenchData { str_cols, hashes: all_hashes, values: all_values, total_rows }
}

// ─── Case A: hashmap only ────────────────────────────────────────────────
#[inline(never)]
fn run_hashmap_only(data: &BenchData, num_chunks: usize) -> usize {
    let mut table = TaperHashMap::with_capacity(num_chunks);
    let mut sum: u64 = 0;
    table.emplace_batch_full(
        &data.hashes,
        &|_: usize, _: &SlotValue| -> bool { true },
        &mut |_i: usize, sv: &mut SlotValue| { sv.bytes = [0x42; 6]; },
        &mut |_i: usize, _sv: &SlotValue, is_new: bool| { if !is_new { sum += 1; } },
    );
    table.len() + sum as usize
}

// ─── Case C: hashmap + NewRow ────────────────────────────────────────────
#[inline(never)]
fn run_hashmap_plus_newrow(data: &BenchData, num_chunks: usize) -> usize {
    let key_sizes = vec![0usize; 4];
    let kinds = vec![ColumnKind::Varchar; 4];
    let mut rc = RowContainer::with_kinds(&key_sizes, &kinds, 8);
    let mut table = TaperHashMap::with_capacity(num_chunks);
    let agg_offset = rc.agg_state_offset();
    let mut sum: u64 = 0;

    let rc_ptr = &mut rc as *mut RowContainer;
    table.emplace_batch_full(
        &data.hashes,
        &|_: usize, _: &SlotValue| -> bool { true },
        &mut |i: usize, sv: &mut SlotValue| {
            let rc = unsafe { &mut *rc_ptr };
            let row = rc.new_row();
            RowContainer::store_value::<i64>(row, agg_offset, data.values[i]);
            sv.set_ptr(row as *const u8);
        },
        &mut |_i: usize, _sv: &SlotValue, is_new: bool| { if !is_new { sum += 1; } },
    );
    table.len() + sum as usize
}

// ─── Case C2: hashmap + NewRow + serialize (no batch compare) ─────────────
#[inline(never)]
fn run_hashmap_plus_serialize(data: &BenchData, num_chunks: usize) -> usize {
    use taper_hashmap::column_marshaller::{serialize_varchar_to_buffer, compute_row_len_size};

    let key_sizes = vec![0usize; 4];
    let kinds = vec![ColumnKind::Varchar; 4];
    let mut rc = RowContainer::with_kinds(&key_sizes, &kinds, 8);
    let mut table = TaperHashMap::with_capacity(num_chunks);
    let agg_offset = rc.agg_state_offset();
    let mut sum: u64 = 0;

    let total_rows = data.hashes.len();
    let num_batches = (total_rows + BATCH_SIZE - 1) / BATCH_SIZE;

    for batch_idx in 0..num_batches {
        let start = batch_idx * BATCH_SIZE;
        let end = (start + BATCH_SIZE).min(total_rows);
        let batch_hashes = &data.hashes[start..end];

        let str_slices: Vec<Vec<&[u8]>> = (0..NUM_STR_COLS)
            .map(|c| data.str_cols[c][start..end].iter().map(|s| s.as_slice()).collect())
            .collect();

        let rc_ptr = &mut rc as *mut RowContainer;

        table.emplace_batch_full(
            batch_hashes,
            &|_: usize, _: &SlotValue| -> bool { true },
            &mut |i: usize, sv: &mut SlotValue| {
                let rc = unsafe { &mut *rc_ptr };
                let row = rc.new_row();
                // Serialize 4 varchars into merged block
                let mut total_size = 0usize;
                for c in 0..NUM_STR_COLS {
                    let s = str_slices[c][i];
                    total_size += 1 + compute_row_len_size(s.len()) as usize + s.len();
                }
                let block = rc.arena_alloc(total_size);
                let mut wp = block;
                for c in 0..NUM_STR_COLS {
                    let s = str_slices[c][i];
                    let written = unsafe { serialize_varchar_to_buffer(wp, s) };
                    wp = unsafe { wp.add(written) };
                }
                // Store pointer at offset 0
                unsafe { (row as *mut *const u8).write_unaligned(block as *const u8); }
                RowContainer::store_value::<i64>(row, agg_offset, data.values[start + i]);
                sv.set_ptr(row as *const u8);
            },
            &mut |_i: usize, _sv: &SlotValue, is_new: bool| { if !is_new { sum += 1; } },
        );
    }
    table.len() + sum as usize
}

// ─── Case D: full pipeline ───────────────────────────────────────────────
#[inline(never)]
fn run_full_pipeline(data: &BenchData, num_chunks: usize) -> usize {
    let mut col_descs: Vec<ColumnDesc> = vec![ColumnDesc::Varchar; NUM_STR_COLS];
    let mut table = TaperColumnSerializeHandler::new(&col_descs, 8, num_chunks);
    let total_rows = data.hashes.len();
    let num_batches = (total_rows + BATCH_SIZE - 1) / BATCH_SIZE;

    for batch_idx in 0..num_batches {
        let start = batch_idx * BATCH_SIZE;
        let end = (start + BATCH_SIZE).min(total_rows);
        let batch_hashes = &data.hashes[start..end];
        let batch_values = &data.values[start..end];
        let str_slices: Vec<Vec<&[u8]>> = (0..NUM_STR_COLS)
            .map(|c| data.str_cols[c][start..end].iter().map(|s| s.as_slice()).collect()).collect();
        let mut columns: Vec<ColumnInput> = Vec::new();
        for c in 0..NUM_STR_COLS { columns.push(ColumnInput::Varchar(&str_slices[c])); }
        table.emplace_table_with_decode(batch_hashes, &columns, batch_values);
    }
    table.num_groups()
}

// ═══════════════════════════════════════════════════════════════════
fn main() {
    let args: Vec<String> = std::env::args().collect();
    let sel: f64 = args.get(1).and_then(|s| s.parse().ok()).unwrap_or(0.1);

    // Second arg: stage filter (A/C/C2/D) or iteration count
    let mut num_iters: usize = 10;
    let mut stage_filter: Option<&str> = None;

    if let Some(arg2) = args.get(2) {
        if arg2.chars().next().map_or(false, |c| c.is_ascii_uppercase()) {
            stage_filter = Some(arg2.as_str());
        } else {
            num_iters = arg2.parse().unwrap_or(10);
        }
    }
    if let Some(arg3) = args.get(3) {
        num_iters = arg3.parse().unwrap_or(10);
    }

    let num_keys = (HT_SIZE as f64 * LOAD_FACTOR) as usize;
    let num_misses = NUM_PROBE_ROWS - (NUM_PROBE_ROWS as f64 * sel) as usize;
    let distinct_keys = num_keys + num_misses;
    let min_slots = ((distinct_keys as f64 / 0.85) as usize).max(8);
    let num_chunks = ((min_slots + 7) / 8).next_power_of_two();

    eprintln!("=== Rust Pipeline Micro Bench ===");
    eprintln!("sel={:.1}, iters={}, totalRows={}, numChunks={}", sel, num_iters, num_keys + NUM_PROBE_ROWS, num_chunks);
    if let Some(s) = stage_filter { eprintln!("Stage filter: {}", s); }
    eprintln!("\nGenerating data...");
    let data = gen_data(num_keys, sel);
    eprintln!("Done.\n");

    let bench = |name: &str, tag: &str, f: fn(&BenchData, usize) -> usize| {
        if let Some(filter) = stage_filter {
            if filter != tag { return; }
        }
        let _ = f(&data, num_chunks); // warmup
        let t0 = Instant::now();
        let mut result = 0;
        for _ in 0..num_iters { result = f(&data, num_chunks); }
        let per_iter = t0.elapsed().as_secs_f64() * 1000.0 / num_iters as f64;
        println!("{:<30}  per_iter={:7.2} ms  result={}", name, per_iter, result);
    };

    println!("=== Results (sel={:.1}, {} iters) ===", sel, num_iters);
    bench("A: hashmap_only", "A", run_hashmap_only);
    bench("C: hashmap+newrow", "C", run_hashmap_plus_newrow);
    bench("C2: hashmap+newrow+serialize", "C2", run_hashmap_plus_serialize);
    bench("D: full_pipeline", "D", run_full_pipeline);
}
